netmaster/k8snetwork/networkpolicy.go

```diff
@@ -391,8 +803,626 @@ func InitK8SServiceWatch(listenURL string, isLeader func() bool) error {
+				TenantName:    np.PodSelector.TenantName,
+				PolicyName:    np.PodSelector.PolicyName,
+				FromIpAddress: fromIp,
+				Priority:      defaultPolicyPriority,
+				Direction:     "in",
+				Action:        "allow"}
+			if isPortsCfg {
+				for _, port := range ingress.IngressRules {
+					//Add port Info into Rule
+					k8sNet.appendPolicyPorts(&rule, port)
+					listRules = append(listRules, rule)
+				}
+			} else {
+				listRules = append(listRules, rule)
+			}
+		}
+	}
+	return &listRules
+}
+
+//final Build Rules by linking from spec to To spec podSelector
+func (k8sNet *k8sContext) finalIngressNetworkPolicyRule(np *k8sNetworkPolicy,
+	toPodIPs []string,
+	ingressRules []client.Rule,
+	isAdd bool) *[]client.Rule {
+
+	var err error
+	ruleList := ingressRules
+
+	//policyCtx := k8sNet.policyRules[np.PodSelector.PolicyName]
+
+	//Ingress Spec To section Pods
+	for _, toIps := range toPodIPs {
+		//npLog.Infof("ruleList:%v", ruleList)
+		//Rebuild Rule List to add To Ips
+		for _, rule := range ruleList {
+			//Update To src Ip section in Rule
+			rule.ToIpAddress = toIps
+			//Generate RuleID :XXX: Should look for better approach
+			rule.RuleID = k8sutils.PolicyToRuleIDUsingIps(
+				toIps, rule.FromIpAddress,
+				rule.Port, rule.Protocol,
+				np.PodSelector.PolicyName)
+
+			npLog.Infof("RulID:%v", rule.RuleID)
+			//Update Policy Name cache with policy Id
+			if isAdd {
+				if err = k8sNet.createRule(&rule); err != nil {
+					npLog.Errorf("failed: rules in-policy %+v, %v",
+						np.PodSelector, err)
+					return nil
+				}
+				//Update RuleID in cache Db
+				//XXX:Should use Hash Set instead of slice
+				// to aviod duplicate Ruleid insertion
+				//policyCtx = append(policyCtx, rule.RuleID)
+			} else { //Policy Delete
+				if err = k8sNet.deleteRule(rule.TenantName,
+					rule.PolicyName,
+					rule.RuleID); err != nil {
+					npLog.Errorf("failed: del in-policy %+v, %v",
+						np.PodSelector, err)
+					return nil
+				}
+				//	for idx, r := range policyCtx {
+				//		if r == rule.RuleID {
+				//			policyCtx = append(policyCtx[0:idx],
+				//				policyCtx[idx+1:]...)
+				//		}
+				//	}
+			}
+			policyMap := k8sNet.policyPerEpg[np.PodSelector.GroupName]
+			policy := policyMap[np.PodSelector.PolicyName]
+			if isAdd {
+				policy = append(policy, rule.RuleID)
+				npLog.Infof("RuleId:%v is added into Policy:%v",
+					rule.RuleID, np.PodSelector.PolicyName)
+			} else {
+				for idx, r := range policy {
+					if r == rule.RuleID {
+						policy = append(policy[0:idx],
+							policy[idx+1:]...)
+						npLog.Infof("RuleId :%v deleted ",
+							rule.RuleID)
+					}
+				}
+			}
+		}
+	}
+	return &ruleList
+}
+
+//Build policy ,rules and attached it to EPG
+func (k8sNet *k8sContext) applyContivNetworkPolicy(np *k8sNetworkPolicy) error {
+	var err error
+	// reset policy to deny on any error
+	policyResetOnErr := func(tenantName, groupName string) {
+		if err != nil {
+			//k8sNet.resetPolicy(tenantName, groupName)
```

Comment:
missing implementation, though this could be touchy, not sure we should back out of the existing working policy . . . we might want to focus more on reconciliation of policy against contiv resources for the next attempt